java/util/src/main/java/com/google/protobuf/util/JsonFormat.java

```diff
@@ -1126,144 +1158,127 @@ public int compare(final Object o1, final Object o2) {
+              stringValue = "-Infinity";
             } else {
-              generator.print("\"Infinity\"");
+              stringValue = "Infinity";
             }
           } else {
-            if (alwaysWithQuotes) {
-              generator.print("\"");
-            }
-            generator.print(doubleValue.toString());
-            if (alwaysWithQuotes) {
-              generator.print("\"");
-            }
+            stringValue = doubleValue.toString();
           }
           break;
 
         case UINT32:
         case FIXED32:
-          if (alwaysWithQuotes) {
-            generator.print("\"");
-          }
-          generator.print(unsignedToString((Integer) value));
-          if (alwaysWithQuotes) {
-            generator.print("\"");
-          }
+          stringValue = unsignedToString((Integer) value);
           break;
 
         case UINT64:
         case FIXED64:
-          generator.print("\"" + unsignedToString((Long) value) + "\"");
+          quoteable = true;
+          stringValue = unsignedToString((Long) value);
           break;
 
         case STRING:
-          generator.print(gson.toJson(value));
+          // `gson.toJson` quotes the value, so we must not quote it again in this method.
+          alwaysWithQuotes = false;
+          stringValue = gson.toJson(value);
           break;
 
         case BYTES:
-          generator.print("\"");
-          generator.print(BaseEncoding.base64().encode(((ByteString) value).toByteArray()));
-          generator.print("\"");
+          alwaysWithQuotes = true;
+          stringValue = BaseEncoding.base64().encode(((ByteString) value).toByteArray());
           break;
 
         case ENUM:
           // Special-case google.protobuf.NullValue (it's an Enum).
           if (field.getEnumType().getFullName().equals("google.protobuf.NullValue")) {
             // No matter what value it contains, we always print it as "null".
-            if (alwaysWithQuotes) {
-              generator.print("\"");
-            }
-            generator.print("null");
-            if (alwaysWithQuotes) {
-              generator.print("\"");
-            }
+            stringValue = "null";
           } else {
             if (printingEnumsAsInts || ((EnumValueDescriptor) value).getIndex() == -1) {
-              generator.print(String.valueOf(((EnumValueDescriptor) value).getNumber()));
+              stringValue = String.valueOf(((EnumValueDescriptor) value).getNumber());
             } else {
-              generator.print("\"" + ((EnumValueDescriptor) value).getName() + "\"");
+              alwaysWithQuotes = true;
+              stringValue = ((EnumValueDescriptor) value).getName();
             }
           }
           break;
 
         case MESSAGE:
         case GROUP:
           print((Message) value);
-          break;
+          return;
+      }
+
+      final boolean shouldQuote = alwaysWithQuotes || (this.alwaysWithQuotes && quoteable);
```

Comment:
This logic statement is a bit hard to read given the local alwaysWithQuotes and the this.alwaysWithQuotes. Seems like it would be better to decide in each case whether to quote or not.